provides.py

```diff
@@ -11,76 +11,65 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from charms.reactive import RelationBase
-from charms.reactive import scopes
-from charms.reactive import hook
-from charms.reactive import not_unless
+from charms.reactive import Endpoint
+from charms.reactive import set_flag, clear_flag
+from charms.reactive import when, when_not
 
 
-class MySQL(RelationBase):
-    # We expect multiple, separate services to be related, but all units of a
-    # given service will share the same database name and connection info.
-    # Thus, we use SERVICE scope and will have one converstaion per service.
-    scope = scopes.SERVICE
+class MySQL(Endpoint):
+    @when('endpoint.{endpoint_name}.joined')
+    def _handle_joined(self):
+        # translate the automatic internal joined flag into a DB requested flag
+        set_flag(self.expand_name('{endpoint_name}.database.requested'))
 
-    @hook('{provides:mysql}-relation-joined')
-    def joined(self):
-        """
-        Handles the relation-joined hook.
-
-        Depending on the state of the conversation, this can trigger the
-        following state:
+    @when_not('endpoint.{endpoint_name}.joined')
+    def _handle_broken(self):
+        # if these were requested but not yet fulfilled, cancel the request
+        clear_flag(self.expand_name('{endpoint_name}.database.requested'))
 
-        * ``{relation_name}.database.requested`` This state will be activated
-          if a remote service is awaiting a database.  This state should be
-          resolved by calling :meth:`provide_database`.  See also
-          :meth:`requested_databases`.
+    def database_requests(self):
         """
-        conversation = self.conversation()
-        conversation.set_state('{relation_name}.database.requested')
+        Return a list of requests for databases.
 
-    @hook('{provides:mysql}-relation-{broken,departed}')
-    def departed(self):
-        conversation = self.conversation()
+        This returns a mapping of request IDs to application names.
 
-        # if these were requested but not yet fulfilled, cancel the request
-        conversation.remove_state('{relation_name}.database.requested')
+        Example usage::
+
+            @when('clients.database.requested')
+            def create_dbs():
+                mysql = endpoint_from_flag('clients.database.requested')
+                for request, application in mysql.database_requests():
```

Comment:
it would be good to add `mysql.database_requests().items()` here.

Reply:
s/good/mandatory since the Python in the example is faulty without the .items() :-)